### PR TITLE
New integration test

### PR DESCRIPTION
**Pull request by corylanou**
Creates an integration test (currently skipped as it fails for a known bug) so we can fix the bug!

### CODE REVIEW

**Comment by corylanou** [comment]
@otoolep This PR is ready for review and merge.  While the test that was the primary driver for this PR is currently skipped, it did flesh out a couple of things.

1) Changing the way we launch the server so that we can test drive it better.
2) Discovered some bugs on unmarshalling the json
3) Added a `Server.Ready()` so we know when the server can accept data.  This needs a closer review by @benbjohnson to know if it is correct or not.  I mostly took an educated guess at what made a server ready.

I'm not sure if the test is actually failing due to a race condition or not, but it doesn't always fail the same way which tells me there are some more race conditions to be addressed yet.

The CLI depends on the JSON bug fixes in this PR, so to take the CLI further, I would like to merge this and come back when we have more stability in some of our race conditions to continue debugging.

File: httpd/handler.go [hunk 1]
```diff
@@ -164,7 +164,7 @@ func (h *Handler) serveWrite(w http.ResponseWriter, r *http.Request, user *influ
 	for {
 		if err := dec.Decode(&br); err != nil {
 			if err.Error() == "EOF" {
-				w.WriteHeader(http.StatusOK)
+				w.WriteHeader(http.StatusCreated)
 				return
```
Comment (otoolep) [inline httpd/handler.go]: Why this change? A resource wasn't created. 201 does not strike me as the correct code.

File: cmd/influxd/server_single_integration_test.go [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+			t.Logf("Failed to clean up %q: %s\n", tmpBrokerDir, err)
+		}
+		err = os.RemoveAll(tmpDataDir)
+		if err != nil {
+			t.Logf("Failed to clean up %q: %s\n", tmpDataDir, err)
+		}
+	}()
+
+	ready := make(chan bool, 1)
+	go func() {
```
Comment (otoolep) [inline cmd/influxd/server_single_integration_test.go]: This has to be in a go-routine or the whole process will block?

File: cmd/influxd/server_single_integration_test.go [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+			case <-time.After(3 * time.Second):
+				if spinupTime == 0 {
+					ellapsed := time.Since(now)
+					t.Fatalf("server failed to spin up in time %v", ellapsed)
+				}
+			}
+		}
+	}()
+
+	// Createa a database
```
Comment (otoolep) [inline cmd/influxd/server_single_integration_test.go]: Nit: typo.

**Comment by otoolep** [comment]
Making sense, though I have some feedback. FWIW, if you have some change here that is blocking something else (the CLI) it's often best practise to simply break that out into its own PR. That helps other devs, since the change will be small, and makes the commit history easier to follow.

File: httpd/handler.go (hunk 1 above)
Comment (corylanou) [inline httpd/handler.go]: If we get EOF, it means that at least one thing was created, otherwise it would of errored out somewhere else in this routine.  This line can only be hit if we did get valid json, wrote it, and now have nothing else to do.  EOF in this case actually means "success".

If anything else in this routine fails, it will exit and write the proper error.

File: cmd/influxd/server_single_integration_test.go (hunk 2 above)
Comment (corylanou) [inline cmd/influxd/server_single_integration_test.go]: Yes, basically this will spin up and start polling, and the following for loop watches the channels to get either the ready state or it times out.  If you have a better pattern than this I'm open to it.

File: httpd/handler.go (hunk 1 above)
Comment (otoolep) [inline httpd/handler.go]: I don't follow you. What was created?

File: cmd/influxd/server_single_integration_test.go (hunk 2 above)
Comment (otoolep) [inline cmd/influxd/server_single_integration_test.go]: That's fine -- as long as you know we need the go-routine.

**Comment by corylanou** [comment]
I don't have a problem with moving the changes to another PR.  I'm not quite sure how to pull them from here though and put them in another one, unless you are suggesting I make them in a different PR, then merge that to master and rebase this branch (which would be easy enough to do).

The other concern I have is that we are moving code really fast, and even 2-3 days can make this branch non-merge-able, and we do need to get these integration tests running.  I'm open to suggestions on how to get this into play sooner than later.

File: httpd/handler.go (hunk 1 above)
Comment (corylanou) [inline httpd/handler.go]: This is the write endpoint, so we actually "created" a entry into our database.  A series/measurement/value was "created" at this point.

Comment (otoolep) [inline httpd/handler.go]: Hmmm. Not a big fan of this -- it's way too low-level a detail, and feels it too pedantic. I think the API should return 200 and be done with it -- I'm thinking of what is natural for client programming. elasticsearch returns 200 when indexing documents in its store.

Comment (otoolep) [inline httpd/handler.go]: In the event a datapoint exists already, our system might overwrite. Will we have a different code for that? I really think the most natural thing for this API to return is 200, and is what client library developers will expect.

Comment (otoolep) [inline httpd/handler.go]: I could be wrong, of course. Any examples of other systems out there that return 204?

**Comment by corylanou** [comment]
@otoolep I don't think this branch is mergable (or will be made mergable).  I'll create a new PR that will create the heartbeat, and one that will create the integration tests.  We should be able to merge both of those easily.

**Comment by corylanou** [comment]
This has been replaced by PR https://github.com/influxdb/influxdb/pull/1490
